common.rkt

```diff
@@ -140,23 +140,35 @@
 (define (reify tm st)
   (define index -1)
   (let ((results (let loop ((tm tm) (st st))
-              (define t (walk tm (state-sub st)))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
-                    (else      st)))))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
     (let* ((walked-sub (walk* tm results))
+           (diseq (sort (state-diseq st) (lambda (x y) (< (length x) (length y)))))
+           (st (diseq-simplify (state (state-sub st) diseq (state-types st))))
+           (diseq (walk* (map pretty-diseq (state-diseq st)) results))
+           (diseq (filter fresh-var-diseq? diseq))
+           (diseq (if (null? diseq) '() (list (cons '=/= diseq))))
            (types (walk* (map pretty-types (state-types st)) results))
-           (diseq (map pretty-diseq (state-diseq st)))
-           (diseq (if (null? diseq) '() (list (walk* (cons '=/= diseq) results))))
+           (types (filter (lambda (t) (not (contains-fresh? (cdr t)))) types))
            (cxs (append types diseq)))
       (if (null? cxs)
           walked-sub
-          (Ans walked-sub cxs)))))
+          (Ans walked-sub (sort cxs term<?))))))
 
 (define (reify/initial-var st)
   (reify initial-var st))
 
+(define (term<? u v)
+  (and (not (eqv? (car u) '=/=)) (eqv? (car v) '=/=)))
+
+(define (contains-fresh? x)
+  (if (pair? x)
+      (or (contains-fresh? (car x)) (contains-fresh? (cdr x)))
+      (var? x)))
+(define (fresh-var-diseq? =/=s) (not (ormap (lambda (=/=) (or (contains-fresh? (car =/=)) (contains-fresh? (cdr =/=)))) =/=s)))
```

Comment:
The name `fresh-var-diseq?` would make more sense if this were answering `#t` when there is a fresh var involved.

Also, it looks like `constains-fresh?` is already recursively checking pair components.  Why do you need to explicitly check the `car` and `cdr` here?

It looks like you could change line 152 to use `(filter-not contains-fresh? diseq)` instead of `(filter fresh-var-diseq? diseq)`.